runtime/compiler/z/codegen/J9S390JNILinkage.cpp

```diff
@@ -0,0 +1,1032 @@
+   // leave a bunch of pinned garbage behind that screws up the gc quality forever
+   TR::LabelSymbol * refPoolRestartLabel = generateLabelSymbol(cg());
+   TR::LabelSymbol * refPoolSnippetLabel = generateLabelSymbol(cg());
+
+   TR::Instruction *gcPoint = generateS390CompareAndBranchInstruction(self()->cg(),
+                                                                      TR::InstOpCode::getCmpOpCode(),
+                                                                      callNode, methodAddressReg,
+                                                                      (signed char)0,
+                                                                      TR::InstOpCode::COND_BNE,
+                                                                      refPoolSnippetLabel, false, true);
+   gcPoint->setNeedsGCMap(0);
+
+   TR::SymbolReference * collapseSymRef = cg()->symRefTab()->findOrCreateRuntimeHelper(TR_S390collapseJNIReferenceFrame,
+                                                                                       false, false, false);
+
+   cg()->addSnippet(new (trHeapMemory()) TR::S390HelperCallSnippet(cg(), callNode,
+      refPoolSnippetLabel, collapseSymRef, refPoolRestartLabel));
+   generateS390LabelInstruction(cg(), TR::InstOpCode::LABEL, callNode, refPoolRestartLabel);
+   }
+
+void
+TR::J9S390JNILinkage::unwrapReturnValue(TR::Node* callNode,
+                                        TR::Register* javaReturnRegister,
+                                        TR::RegisterDependencyConditions* deps)
+   {
+   TR::LabelSymbol * tempLabel = generateLabelSymbol(cg());
+   TR::Instruction* start = generateS390CompareAndBranchInstruction(cg(), TR::InstOpCode::getCmpOpCode(),
+                                                                    callNode, javaReturnRegister,
+                                                                    (signed char)0, TR::InstOpCode::COND_BE, tempLabel);
+
+   start->setStartInternalControlFlow();
+   generateRXInstruction(cg(), TR::InstOpCode::getLoadOpCode(), callNode, javaReturnRegister,
+              generateS390MemoryReference(javaReturnRegister, 0, cg()));
+
+   int32_t regPos = deps->searchPostConditionRegisterPos(javaReturnRegister);
+   TR::RealRegister::RegNum realReg = deps->getPostConditions()->getRegisterDependency(regPos)->getRealRegister();
+   TR::RegisterDependencyConditions * postDeps = new (trHeapMemory()) TR::RegisterDependencyConditions(0, 1, cg());
+   postDeps->addPostCondition(javaReturnRegister, realReg);
+
+   TR::Instruction* end =generateS390LabelInstruction(cg(), TR::InstOpCode::LABEL, callNode, tempLabel);
+   end->setEndInternalControlFlow();
+   end->setDependencyConditions(postDeps);
+   }
+
+extern "C" void fast_jitPreJNICallOffloadCheck();
+extern "C" void fast_jitPostJNICallOffloadCheck();
+
+void
+TR::J9S390JNILinkage::callIFAOffloadCheck(bool isPreJNIHelper, bool isJNICallOutFrame)
```

Comment:
We should fold this function away. I don't think it's needed. The environment is already present via `TR_S390CEnvironmentAddress` so there is no need to deal with the function pointers etc. which renders this function an effective three liner:

```
TR::SymbolReference * offloadSymRef = cg()->symRefTab()->findOrCreateRuntimeHelper(helperIndex, false, false, false);
TR::Node* helperCallNode = TR::Node::createWithSymRef(TR::call, 0, offloadSymRef);
cHelperLinkage->buildDirectDispatch(helperCallNode, NULL, NULL, true,
                                    isJNICallOutFrame,
                                    reinterpret_cast<intptrj_t>(TR_S390CEnvironmentAddress));
```

I'm not sure it has a need for existence.